### PR TITLE
Fix correlation of series with same names

### PR DESCRIPTION
**Pull request by Chilipp**
closes https://github.com/dask/dask/issues/4906

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks. Do you have time to push an update with a small test that tests the original problem?

**Comment by Chilipp** [comment]
Sure! See https://github.com/dask/dask/pull/4934/commits/c8248dd6545f376b6bb248b05cc3b59be644328e where I just duplicated the current test_corr for series correlations

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -2434,6 +2434,27 @@ def test_corr():
...
+    db = dd.from_pandas(b, npartitions=7)
+
+    res = da.corr(db)
+    res2 = da.corr(db, split_every=2)
+    res3 = da.corr(db, min_periods=10)
+    res4 = da.corr(db, min_periods=10, split_every=2)
+    sol = da.corr(db)
+    sol2 = da.corr(db, min_periods=10)
+    assert_eq(res, sol)
+    assert_eq(res2, sol)
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_dataframe.py]: Seems this is failing CI.

I think a simpler test will suffice here. Something basic like

```
result = ddf.A.corr(ddf.A)
expected = df.A.corr(df.A)
assert_eq(result, expected)
```

Comment (Chilipp) [reply dask/dataframe/tests/test_dataframe.py]: True. But nevertheless: any idea why it fails? I tried it locally and the values disagree by something around 0.4!

**Comment by TomAugspurger** [comment]
Oh, should it be `assert_eq(res2, sol2)` instead of `sol`?

**Comment by Chilipp** [comment]
Hmm, I'm afraid, it doesn't. And in fact, I just literally copied the original test and just changed `df.B` into `df.B.rename('A')`. 

You can also see with the further test implemented in https://github.com/dask/dask/pull/4934/commits/d27b54c747e7144813397ebc0949baac223d7899 that it changes a lot if I rename the column and the results do not agree anymore. I can try to dig deeper into this tomorrow, but so far, I have no idea where this comes from.

**Comment by Chilipp** [comment]
Alright @TomAugspurger,  I found the issue and solved it in https://github.com/dask/dask/pull/4934/commits/8db2cb4595bd12bcca6f6638f132a8f660aed6c0. It was another access to `df[col]` that fails if there are duplicated column names. It should be solved now.

Do you want me to reduce the number of tests that I implemented in https://github.com/dask/dask/pull/4934/commits/d27b54c747e7144813397ebc0949baac223d7899 and https://github.com/dask/dask/pull/4934/commits/c8248dd6545f376b6bb248b05cc3b59be644328e?

**Comment by jcrist** [comment]
@Chilipp, yes it would be good to reduce the length of your added tests to something minimal that would catch this error. This makes our test suite faster, and our test code easier to understand (for future maintenance).

**Comment by Chilipp** [comment]
> yes it would be good to reduce the length of your added tests

alright @jcrist, thanks for the feedback! This has been implemented in  https://github.com/dask/dask/pull/4934/commits/40cf5cb84c3514c6fffeb0a44a5c38e1acf76341 as a separate but shorter `test_corr_same_name` test function

File: dask/dataframe/tests/test_dataframe.py [hunk 2]
```diff
@@ -2436,6 +2436,20 @@ def test_corr():
 
     pytest.raises(NotImplementedError, lambda: da.corr(db, method='spearman'))
     pytest.raises(TypeError, lambda: da.corr(ddf))
+    
+def test_corr_same_name():
```
Comment (jcrist) [inline dask/dataframe/tests/test_dataframe.py]: You'll need a newline above here to pass the flake8 tests. You can check your code passes beforehand as:

```
$ flake8 dask
```

Comment (Chilipp) [reply dask/dataframe/tests/test_dataframe.py]: my bad, sorry. I thought I can quickly change it through the Github API but it's too prone to typos. I fixed it in https://github.com/dask/dask/pull/4934/commits/06dc7cccfdb18f82b7b503440ecaedaa6e7541ec

**Comment by TomAugspurger** [comment]
@Chilipp can you clarify https://github.com/dask/dask/pull/4934#issuecomment-502060839 a bit? Is the issue that we don't correctly handle duplicate columns on master? Or does your PR not correctly handle duplicates?

**Comment by Chilipp** [comment]
> Is the issue that we don't correctly handle duplicate columns on master? Or does your PR not correctly handle duplicates?

The problem is the current implementation. It iterates over the columns and accesses it using the `__getitem__` method:

```python
for idx, col in enumerate(df):
    mask = df[col].notnull()
```
The `col` however appears twice in the dataframe columns, and therefore `mask` is an array of shape `(N, 2)`. I did not see this immediately, because, in contrast to what has been fixed by https://github.com/dask/dask/pull/4934/commits/453162caa99f303be9e430d4f3e9509d21fe1657, it did not raise an error but just changed the result. Additionally, and I am not entirely sure why, it did only provide wrong results if `split_every` is set. Nevertheless, I fixed it now and the tests implemented in https://github.com/dask/dask/pull/4934/commits/06dc7cccfdb18f82b7b503440ecaedaa6e7541ec show that it works.

I think the important message from this issue is that you should not use the column names to access the series in a dataframe but rather the index of the column.

**Comment by TomAugspurger** [comment]
Yep, thanks for confirming.

**Pull request merged**